File: next/src/shared/utils/pageRanges.ts
import type { PdfSplitRule } from '../contracts/pdf'

export function parsePageSelection(expression: string, maxPage: number): number[] {
  if (!Number.isSafeInteger(maxPage) || maxPage < 1) throw new Error('PDF has no pages')
  const value = expression.trim().replace(/[，,]/g, ';').replace(/\s+/g, '')
  if (!value || !/^\d+(?:-\d+)?(?:;\d+(?:-\d+)?)*$/.test(value)) throw new Error('Use page ranges such as 1-5;8;10-12')
  const pages = new Set<number>()
  for (const token of value.split(';')) {
    const [startText, endText = startText] = token.split('-')
    const start = Number(startText)
    const end = Number(endText)
    if (start < 1 || end < start || end > maxPage) throw new Error(`Page range must stay between 1 and ${maxPage}`)
    for (let page = start; page <= end; page += 1) pages.add(page)
  }
  return [...pages]
}

export function selectSplitPages(pageRange: string, rule: PdfSplitRule, customRule: string, maxPage: number): number[] {
  const candidates = parsePageSelection(pageRange, maxPage)
  if (rule === 'odd') return candidates.filter((page) => page % 2 === 1)
  if (rule === 'even') return candidates.filter((page) => page % 2 === 0)
  const selected = new Set(parsePageSelection(customRule, maxPage))
  return candidates.filter((page) => selected.has(page))
}

File: next-flutter/lib/features/pdf/page_ranges.dart
List<int> parsePageSelection(String expression, int maxPage) {
  if (maxPage < 1) throw const FormatException('PDF has no pages');
  final value =
      expression.trim().replaceAll(RegExp('[，,]'), ';').replaceAll(RegExp(r'\s+'), '');
  if (value.isEmpty ||
      !RegExp(r'^\d+(?:-\d+)?(?:;\d+(?:-\d+)?)*$').hasMatch(value)) {
    throw const FormatException('Use page ranges such as 1-5;8;10-12');
  }
  final pages = <int>{};
  for (final token in value.split(';')) {
    final parts = token.split('-');
    final start = int.parse(parts[0]);
    final end = int.parse(parts.length > 1 ? parts[1] : parts[0]);
    if (start < 1 || end < start || end > maxPage) {
      throw FormatException('Page range must stay between 1 and $maxPage');
    }
    for (var page = start; page <= end; page += 1) {
      pages.add(page);
    }
  }
  return pages.toList();
}

List<int> selectSplitPages(
    String pageRange, String rule, String customRule, int maxPage) {
  final candidates = parsePageSelection(pageRange, maxPage);
  if (rule == 'odd') {
    return [for (final page in candidates) if (page.isOdd) page];
  }
  if (rule == 'even') {
    return [for (final page in candidates) if (page.isEven) page];
  }
  final selected = parsePageSelection(customRule, maxPage).toSet();
  return [for (final page in candidates) if (selected.contains(page)) page];
}
